internal/overlord/logstate/manager.go

```diff
@@ -0,0 +1,185 @@
+				iterator = m.buffers[serviceName].HeadIterator(0)
+			}
+
+			// Copy iterator reference over
+			iterators[serviceName][*target] = iterator
+
+			m.startForwarder(serviceName, targetName, target, iterator)
+		}
+	}
+
+	// For other services: forwarders will be started in ServiceStarted
+
+	// Update buffers and iterators
+	m.buffers = buffers
+	m.iterators = iterators
+
+	// Update services
+	m.services = map[string]*plan.Service{}
+	for serviceName, service := range pl.Services {
+		m.services[serviceName] = service.Copy()
+	}
+}
+
+// selectTargets goes through the given plan and determines which service's
+// logs should be forwarded to which log targets.
+func selectTargets(p *plan.Plan) map[string]map[string]*plan.LogTarget {
+	planTargets := make(map[string]map[string]*plan.LogTarget, len(p.Services))
+
+	for serviceName, service := range p.Services {
+		planTargets[serviceName] = make(map[string]*plan.LogTarget, len(service.LogTargets))
+
+		for _, targetName := range service.LogTargets {
+			target, ok := p.LogTargets[targetName]
+			if !ok {
+				logger.Noticef("Internal error: log manager cannot find target %q (specified by service %q in plan)",
+					targetName, serviceName)
+				continue
+			}
+			if target.Selection != plan.DisabledSelection {
+				planTargets[serviceName][targetName] = target
+			}
+		}
+
+		// Add opt-out targets if none specified
+		if len(service.LogTargets) == 0 {
+			for targetName, target := range p.LogTargets {
+				// Default (unset) is opt-out
+				if target.Selection == plan.OptOutSelection || target.Selection == plan.UnsetSelection {
+					planTargets[serviceName][targetName] = target
+				}
+			}
+		}
+	}
+
+	return planTargets
+}
+
+// ServiceStarted notifies the log manager that the named service has started,
+// and provides a reference to the service's log buffer.
+func (m *LogManager) ServiceStarted(serviceName string, buffer *servicelog.RingBuffer) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.buffers[serviceName] = buffer
+	m.iterators[serviceName] = make(map[plan.LogTarget]servicelog.Iterator,
+		len(m.planTargets[serviceName]))
+
+	for targetName, target := range m.planTargets[serviceName] {
+		iterator := m.buffers[serviceName].HeadIterator(0)
+		m.iterators[serviceName][*target] = iterator
+
+		m.startForwarder(serviceName, targetName, target, iterator)
+	}
+}
+
+// startForwarder starts a new forwarder for the specified service and target,
+// and adds this forwarder to the LogManager's list.
+func (m *LogManager) startForwarder(
+	serviceName, targetName string, target *plan.LogTarget, iterator servicelog.Iterator,
+) {
+	f, err := newLogForwarder(serviceName, targetName, target, iterator)
+	if err != nil {
+		logger.Noticef("Internal error: cannot create forwarder %q -> %q: %v",
+			serviceName, targetName, err)
+		return
+	}
+
+	m.forwarders = append(m.forwarders, f)
+	logger.Debugf("start forwarding logs for service %q to target %q",
+		serviceName, targetName)
+	go f.forward()
```

Comment:
Do we want to have a "log manager shutdown" method (I seem to recall Stop or Shutdown or something is called automatically on managers), to ensure graceful shutdown when Pebble itself is stopped?

Reply:
yes, then the `LogManager` would have implement the following interfaces in the `overlord` package:
https://github.com/canonical/pebble/blob/054e5fbebc6bb0b1b195c709baa321c6c7aa2501/internal/overlord/stateengine.go#L25-L31
https://github.com/canonical/pebble/blob/054e5fbebc6bb0b1b195c709baa321c6c7aa2501/internal/overlord/stateengine.go#L40-L47

Reply:
Just the `Stop()` one, right?

Reply:
nah, we'll need to implement both. StateStopper is only used on top of StateManager